File: src/xyang/ast.py
"""
Abstract Syntax Tree (AST) nodes for YANG statements.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Callable, List, Optional
from dataclasses import dataclass, field

from .errors import YangSemanticError

if TYPE_CHECKING:
    from .xpath.ast import ASTNode, PathNode
    from .module import YangModule


def _empty_path_node():
    from .xpath.ast import PathNode

    return PathNode([])


@dataclass
class YangStatementList:
    """Base for nodes that contain a list of YANG statements (module body or statement body)."""
    statements: List['YangStatement'] = field(default_factory=list)

    def find_statement(self, name: str) -> Optional['YangStatement']:
        """Find a statement by name."""
        for stmt in self.statements:
            if stmt.name == name:
                return stmt
        return None

    def get_all_leaves(self) -> List['YangLeafStmt']:
        """Get all leaf statements recursively."""
        leaves = []
        for stmt in self.statements:
            leaves.extend(self._collect_leaves(stmt))
        return leaves

    def _collect_leaves(self, stmt: 'YangStatement') -> List['YangLeafStmt']:
        """Recursively collect leaf statements."""
        leaves = []
        if isinstance(stmt, YangLeafStmt):
            leaves.append(stmt)
        elif isinstance(stmt, (YangContainerStmt, YangListStmt)):
            for child in stmt.statements:
                leaves.extend(self._collect_leaves(child))
        return leaves


@dataclass
class YangStatement(YangStatementList):
    """Base class for all YANG statements."""
    name: str = ""
    description: str = ""

    def get_schema_node(self) -> Optional[str]:
        """Path segment for this statement in the schema tree, or None if it adds no node.

        Used when assembling paths for ``refine`` targets and similar (e.g. ``uses`` is
        not a schema node; see :class:`YangUsesStmt`).
        """
        return None

    def child_names(self, data: dict) -> set[str]:
        return {self.name} if getattr(self, "name", None) else set()


@dataclass
class YangStatementWithMust(YangStatement):
    """Statement that can have must constraints (leaf, leaf-list, container, list)."""
    must_statements: List['YangMustStmt'] = field(default_factory=list)


@dataclass
class YangStatementWithWhen(YangStatement):
    """Statement that can have a when condition (leaf, leaf-list, container, list)."""

    when: Optional['YangWhenStmt'] = None
    # RFC 7950 / 7952: one or more ``if-feature`` substatements (AND of expressions).
    if_features: List[str] = field(default_factory=list)

    def get_schema_node(self) -> Optional[str]:
        """Default for data-shaped nodes: instance key is ``name``. Override when not a data node."""
        return self.name or None


@dataclass
class YangTypedefStmt(YangStatement):
    """Typedef statement."""
    type: Optional['YangTypeStmt'] = None

    def get_schema_node(self) -> Optional[str]:
        return self.name or None


@dataclass
class YangIdentityStmt(YangStatement):
    """Identity statement (RFC 7950); may have multiple ``base`` substatements (YANG 1.1)."""

    bases: List[str] = field(default_factory=list)
    if_features: List[str] = field(default_factory=list)

    def get_schema_node(self) -> Optional[str]:
        return None


@dataclass
class YangBitStmt:
    """Single bit in a ``type bits { ... }`` (RFC 7950 §9.3.4).

    ``position`` is ``None`` until the parser assigns implicit positions; after
    parsing a complete ``bits`` type, every bit has a non-negative integer position.
    """

    name: str
    position: Optional[int] = None


@dataclass
class YangPatternSpec:
    """One ``pattern`` substatement on ``type string`` (RFC 7950 §9.4.6)."""

    pattern: str
    invert_match: bool = False
    error_message: Optional[str] = None
    error_app_tag: Optional[str] = None


@dataclass
class YangTypeStmt:
    """Type statement.

    For ``pattern`` under string-related types (RFC 7950 §9.4.6), ``patterns``
    preserves all occurrences and optional substatements (including
    ``modifier invert-match`` in YANG 1.1). Legacy flat fields
    ``pattern`` / ``pattern_error_message`` / ``pattern_error_app_tag`` are
    kept for compatibility and mirror the last pattern entry when present.
    """

    name: str
    pattern: Optional[str] = None
    pattern_error_message: Optional[str] = None
    pattern_error_app_tag: Optional[str] = None
    patterns: List[YangPatternSpec] = field(default_factory=list)
    length: Optional[str] = None
    range: Optional[str] = None
    fraction_digits: Optional[int] = None
    enums: List[str] = field(default_factory=list)
    bits: List[YangBitStmt] = field(default_factory=list)  # For bits type
    types: List['YangTypeStmt'] = field(default_factory=list)  # For union types
    path: Optional["PathNode"] = None  # For leafref: parsed XPath path, set during parsing
    require_instance: bool = True  # For leafref
    identityref_bases: List[str] = field(default_factory=list)  # For identityref: one per ``base`` substatement


@dataclass
class YangContainerStmt(YangStatementWithMust, YangStatementWithWhen):
    """Container statement."""
    presence: Optional[str] = None


@dataclass
class YangListStmt(YangStatementWithMust, YangStatementWithWhen):
    """List statement."""
    key: Optional[str] = None
    min_elements: Optional[int] = None
    max_elements: Optional[int] = None


@dataclass
class YangLeafStmt(YangStatementWithMust, YangStatementWithWhen):
    """Leaf statement."""
    type: Optional[YangTypeStmt] = None
    mandatory: bool = False
    default: Optional[Any] = None


@dataclass
class YangLeafListStmt(YangStatementWithMust, YangStatementWithWhen):
    """Leaf-list statement."""

    type: Optional[YangTypeStmt] = None
    min_elements: Optional[int] = None
    max_elements: Optional[int] = None
    # RFC 7950: 0..n ``default`` substatements (ordered set of default values); filled by refine too
    defaults: List[Any] = field(default_factory=list)


@dataclass
class YangAnydataStmt(YangStatementWithMust, YangStatementWithWhen):
    """Anydata statement (RFC 7950 §7.12); instance is an arbitrary JSON-compatible value."""

    mandatory: bool = False


@dataclass
class YangAnyxmlStmt(YangStatementWithMust, YangStatementWithWhen):
    """Anyxml statement (RFC 7950 §7.11, deprecated in YANG 1.1); treated like anydata for JSON."""

    mandatory: bool = False


@dataclass
class YangExtensionStmt(YangStatement):
    """YANG ``extension`` definition (RFC 7950 §7.17)."""

    argument_name: str = ""
    argument_yin_element: Optional[bool] = None
    apply_callback: Optional[
        Callable[["YangExtensionInvocationStmt", "YangModule"], Optional["YangStatement"]]
    ] = None

    def apply(
        self,
        invocation: "YangExtensionInvocationStmt",
        *,
        context_module: "YangModule",
    ) -> Optional["YangStatement"]:
        if self.apply_callback is None:
            return invocation
        return self.apply_callback(invocation, context_module)

    def get_schema_node(self) -> Optional[str]:
        return None


@dataclass
class YangExtensionInvocationStmt(YangStatementWithMust, YangStatementWithWhen):
    """A generic prefixed extension invocation (e.g. ``abc:foo ...``)."""

    prefix: str = ""
    resolved_module: "YangModule" = field(default=None)  # type: ignore[assignment]
    resolved_extension: YangExtensionStmt = field(default=None)  # type: ignore[assignment]
    argument: Optional[str] = None

    def __post_init__(self) -> None:
        if not self.prefix:
            raise ValueError("extension invocation requires a non-empty prefix")
        if self.resolved_module is None:
            raise ValueError("extension invocation requires resolved_module")
        if self.resolved_extension is None:
            raise ValueError("extension invocation requires resolved_extension")

    def get_schema_node(self) -> Optional[str]:
        return None



@dataclass
class YangParsedXPathBase:
    """Base class for statements that carry a parsed XPath expression."""
    expression: str
    description: str = ""

    ast: ASTNode | None = None  # Parsed XPath AST for reuse; always set in __post_init__

    def __post_init__(self) -> None:
        # Lazily import to avoid circular dependencies at module import time.
        if self.expression:
            from .xpath import XPathParser
            self.ast = XPathParser(self.expression).parse()

@dataclass
class YangMustStmt(YangParsedXPathBase):
    """Must statement (constraint)."""
    error_message: str = ""


@dataclass
class YangWhenStmt(YangParsedXPathBase):
    """When statement (conditional). XPath lives in ``expression`` (see ``condition``)."""

    # RFC 7950 §7.21.5: ``when`` under ``uses`` uses the parent of ``uses`` as context node.
    # Set when this ``when`` is merged from a ``uses`` (or AND-merged with such a ``when``).
    evaluate_with_parent_context: bool = False

    @property
    def condition(self) -> str:
        return self.expression


@dataclass
class YangLeafrefStmt:
    """Leafref type statement."""
    path: str = ""
    require_instance: bool = True


@dataclass
class YangGroupingStmt(YangStatement):
    """Grouping statement - defines reusable schema components."""
    pass


@dataclass
class YangUsesStmt(YangStatementWithWhen):
    """Uses statement - incorporates a grouping."""

    grouping_name: str = ""
    refines: List['YangRefineStmt'] = field(default_factory=list)

    def get_schema_node(self) -> Optional[str]:
        return None


@dataclass
class YangAugmentStmt(YangStatementWithWhen):
    """Augment statement (RFC 7950). After parse, children are merged into the target node (see ``augment_expand``)."""

    augment_path: str = ""

    def get_schema_node(self) -> Optional[str]:
        return None


@dataclass
class YangRefineStmt(YangStatementWithMust):
    """Refine statement - modifies nodes from a grouping when using it."""
    target_path: "PathNode" = field(default_factory=_empty_path_node)  # Descendant path (parsed; use .to_string() for display)
    type: Optional['YangTypeStmt'] = None  # Refined type when target is a leaf
    min_elements: Optional[int] = None  # Refined min-elements (list / leaf-list)
    max_elements: Optional[int] = None  # Refined max-elements (list / leaf-list)
    # RFC 7950 §7.13.2: ``default`` substatements (one value each); one for leaf targets, 1..n for leaf-list
    refined_defaults: List[Any] = field(default_factory=list)
    # RFC 7950 §7.13.2: leaf / choice may get a different mandatory; None = omit from refine
    refined_mandatory: Optional[bool] = None
    # RFC 7950 §7.13.2: additional if-feature expressions (AND with target node's own).
    if_features: List[str] = field(default_factory=list)


@dataclass
class YangChoiceStmt(YangStatementWithWhen):
    """Choice statement - defines mutually exclusive alternatives."""

    mandatory: bool = False
    cases: List['YangCaseStmt'] = field(default_factory=list)

    def child_names(self, data: dict) -> set[str]:
        for case in self.cases:
            if any(getattr(s, "name", None) in data for s in case.statements):
                return {s.name for s in case.statements if getattr(s, "name", None)}
        return set()

    def validate_case_unique_child_names(self) -> None:
        """RFC 7950 §7.9: case branches share one namespace; schema node names must be unique."""
        seen: dict[str, str] = {}
        for case in self.cases:
            for sub in case.statements:
                seg = sub.get_schema_node()
                if seg is None:
                    continue
                if seg in seen:
                    prev_case = seen[seg]
                    raise YangSemanticError(
                        f"Choice {self.name!r}: schema node {seg!r} appears in case "
                        f"{prev_case!r} and again in case {case.name!r} "
                        "(RFC 7950: names of nodes in the cases of a choice must be unique)."
                    )
                seen[seg] = case.name


@dataclass
class YangCaseStmt(YangStatementWithWhen):
    """Case statement - defines one alternative in a choice."""

    def child_names(self, data: dict) -> set[str]:
        return {s.name for s in self.statements if getattr(s, "name", None)}

File: ts/src/core/ast.ts
import { YangSemanticError } from "./errors";
import { parseXPath } from "../xpath/parser";
import type { XPathAstNode } from "../xpath/ast";
import type { YangModule } from "./module";

export type ASTNode = XPathAstNode;
export type PathNode = XPathAstNode;

export class YangStatementList {
  statements: YangStatement[];

  constructor(statements: YangStatement[] = []) {
    this.statements = statements;
  }

  find_statement(name: string): YangStatement | undefined {
    return this.statements.find((stmt) => stmt.name === name);
  }

  findStatement(name: string): YangStatement | undefined {
    return this.find_statement(name);
  }

  get_all_leaves(): YangLeafStmt[] {
    const leaves: YangLeafStmt[] = [];
    for (const stmt of this.statements) {
      leaves.push(...this.collectLeaves(stmt));
    }
    return leaves;
  }

  getAllLeaves(): YangLeafStmt[] {
    return this.get_all_leaves();
  }

  private collectLeaves(stmt: YangStatement): YangLeafStmt[] {
    if (stmt instanceof YangLeafStmt) {
      return [stmt];
    }
    if (stmt instanceof YangContainerStmt || stmt instanceof YangListStmt) {
      const leaves: YangLeafStmt[] = [];
      for (const child of stmt.statements) {
        leaves.push(...this.collectLeaves(child));
      }
      return leaves;
    }
    return [];
  }
}

export class YangStatement extends YangStatementList {
  keyword: string;
  name: string;
  description: string;

  constructor(init: Partial<YangStatement> = {}) {
    super(init.statements ?? []);
    this.keyword = init.keyword ?? "";
    this.name = init.name ?? "";
    this.description = init.description ?? "";
  }

  get_schema_node(): string | undefined {
    return undefined;
  }

  getSchemaNode(): string | undefined {
    return this.get_schema_node();
  }

  child_names(_data: Record<string, unknown>): Set<string> {
    return this.name ? new Set([this.name]) : new Set();
  }

  childNames(data: Record<string, unknown>): Set<string> {
    return this.child_names(data);
  }
}

export class YangStatementWithMust extends YangStatement {
  must_statements: YangMustStmt[];

  constructor(init: Partial<YangStatementWithMust> = {}) {
    super(init);
    this.must_statements = init.must_statements ?? [];
  }
}

export class YangStatementWithWhen extends YangStatement {
  when?: YangWhenStmt;
  if_features: string[];

  constructor(init: Partial<YangStatementWithWhen> = {}) {
    super(init);
    this.when = init.when;
    this.if_features = init.if_features ?? [];
  }

  override get_schema_node(): string | undefined {
    return this.name || undefined;
  }
}

export class YangTypedefStmt extends YangStatement {
  type?: YangTypeStmt;

  constructor(init: Partial<YangTypedefStmt> = {}) {
    super(init);
    this.keyword = "typedef";
    this.type = init.type;
  }

  override get_schema_node(): string | undefined {
    return this.name || undefined;
  }
}

export class YangIdentityStmt extends YangStatement {
  bases: string[];
  if_features: string[];

  constructor(init: Partial<YangIdentityStmt> = {}) {
    super(init);
    this.keyword = "identity";
    this.bases = init.bases ?? [];
    this.if_features = init.if_features ?? [];
  }

  override get_schema_node(): string | undefined {
    return undefined;
  }
}

export class YangBitStmt {
  name: string;
  position?: number;

  constructor(init: Partial<YangBitStmt> = {}) {
    this.name = init.name ?? "";
    this.position = init.position;
  }
}

export class YangTypeStmt {
  name: string;
  pattern?: string;
  pattern_error_message?: string;
  pattern_error_app_tag?: string;
  length?: string;
  range?: string;
  fraction_digits?: number;
  enums: string[];
  bits: YangBitStmt[];
  types: YangTypeStmt[];
  path?: PathNode;
  require_instance: boolean;
  identityref_bases: string[];

  constructor(init: Partial<YangTypeStmt> = {}) {
    this.name = init.name ?? "";
    this.pattern = init.pattern;
    this.pattern_error_message = init.pattern_error_message;
    this.pattern_error_app_tag = init.pattern_error_app_tag;
    this.length = init.length;
    this.range = init.range;
    this.fraction_digits = init.fraction_digits;
    this.enums = init.enums ?? [];
    this.bits = init.bits ?? [];
    this.types = init.types ?? [];
    this.path = init.path;
    this.require_instance = init.require_instance ?? true;
    this.identityref_bases = init.identityref_bases ?? [];
  }
}

export class YangContainerStmt extends YangStatementWithWhen {
  must_statements: YangMustStmt[];
  presence?: string;

  constructor(init: Partial<YangContainerStmt> = {}) {
    super(init);
    this.keyword = "container";
    this.must_statements = init.must_statements ?? [];
    this.presence = init.presence;
  }
}

export class YangListStmt extends YangStatementWithWhen {
  must_statements: YangMustStmt[];
  key?: string;
  min_elements?: number;
  max_elements?: number;

  constructor(init: Partial<YangListStmt> = {}) {
    super(init);
    this.keyword = "list";
    this.must_statements = init.must_statements ?? [];
    this.key = init.key;
    this.min_elements = init.min_elements;
    this.max_elements = init.max_elements;
  }
}

export class YangLeafStmt extends YangStatementWithWhen {
  must_statements: YangMustStmt[];
  type?: YangTypeStmt;
  mandatory: boolean;
  default?: unknown;

  constructor(init: Partial<YangLeafStmt> = {}) {
    super(init);
    this.keyword = "leaf";
    this.must_statements = init.must_statements ?? [];
    this.type = init.type;
    this.mandatory = init.mandatory ?? false;
    this.default = init.default;
  }
}

export class YangLeafListStmt extends YangStatementWithWhen {
  must_statements: YangMustStmt[];
  type?: YangTypeStmt;
  min_elements?: number;
  max_elements?: number;
  defaults: unknown[];

  constructor(init: Partial<YangLeafListStmt> = {}) {
    super(init);
    this.keyword = "leaf-list";
    this.must_statements = init.must_statements ?? [];
    this.type = init.type;
    this.min_elements = init.min_elements;
    this.max_elements = init.max_elements;
    this.defaults = init.defaults ?? [];
  }
}

export class YangAnydataStmt extends YangStatementWithWhen {
  must_statements: YangMustStmt[];
  mandatory: boolean;

  constructor(init: Partial<YangAnydataStmt> = {}) {
    super(init);
    this.keyword = "anydata";
    this.must_statements = init.must_statements ?? [];
    this.mandatory = init.mandatory ?? false;
  }
}

export class YangAnyxmlStmt extends YangStatementWithWhen {
  must_statements: YangMustStmt[];
  mandatory: boolean;

  constructor(init: Partial<YangAnyxmlStmt> = {}) {
    super(init);
    this.keyword = "anyxml";
    this.must_statements = init.must_statements ?? [];
    this.mandatory = init.mandatory ?? false;
  }
}

export type ExtensionApplyCallback = (
  invocation: YangExtensionInvocationStmt,
  contextModule: YangModule
) => YangStatement | undefined;

export class YangExtensionStmt extends YangStatement {
  argument_name: string;
  argument_yin_element?: boolean;
  apply_callback?: ExtensionApplyCallback;

  constructor(init: Partial<YangExtensionStmt> = {}) {
    super(init);
    this.keyword = "extension";
    this.argument_name = init.argument_name ?? "";
    this.argument_yin_element = init.argument_yin_element;
    this.apply_callback = init.apply_callback;
  }

  apply(invocation: YangExtensionInvocationStmt, options: { context_module: YangModule }): YangStatement | undefined {
    if (!this.apply_callback) {
      return invocation;
    }
    return this.apply_callback(invocation, options.context_module);
  }

  override get_schema_node(): string | undefined {
    return undefined;
  }
}

export class YangExtensionInvocationStmt extends YangStatementWithWhen {
  must_statements: YangMustStmt[];
  prefix: string;
  resolved_module: YangModule;
  resolved_extension: YangExtensionStmt;
  argument?: string;

  constructor(init: {
    prefix: string;
    resolved_module: YangModule;
    resolved_extension: YangExtensionStmt;
    argument?: string;
  } & Partial<YangExtensionInvocationStmt>) {
    super(init);
    this.keyword = "extension-invocation";
    this.must_statements = init.must_statements ?? [];
    this.prefix = init.prefix;
    this.resolved_module = init.resolved_module;
    this.resolved_extension = init.resolved_extension;
    this.argument = init.argument;

    if (!this.prefix) {
      throw new Error("extension invocation requires a non-empty prefix");
    }
    if (!this.resolved_module) {
      throw new Error("extension invocation requires resolved_module");
    }
    if (!this.resolved_extension) {
      throw new Error("extension invocation requires resolved_extension");
    }
  }

  override get_schema_node(): string | undefined {
    return undefined;
  }
}

export class YangParsedXPathBase {
  expression: string;
  description: string;
  ast?: ASTNode;

  constructor(init: { expression: string; description?: string }) {
    this.expression = init.expression;
    this.description = init.description ?? "";
    if (this.expression) {
      this.ast = parseXPath(this.expression);
    }
  }
}

export class YangMustStmt extends YangParsedXPathBase {
  error_message: string;

  constructor(init: { expression: string; description?: string; error_message?: string }) {
    super(init);
    this.error_message = init.error_message ?? "";
  }
}

export class YangWhenStmt extends YangParsedXPathBase {
  evaluate_with_parent_context: boolean;

  constructor(init: { expression: string; description?: string; evaluate_with_parent_context?: boolean }) {
    super(init);
    this.evaluate_with_parent_context = init.evaluate_with_parent_context ?? false;
  }

  get condition(): string {
    return this.expression;
  }
}

export class YangLeafrefStmt {
  path: string;
  require_instance: boolean;

  constructor(init: Partial<YangLeafrefStmt> = {}) {
    this.path = init.path ?? "";
    this.require_instance = init.require_instance ?? true;
  }
}

export class YangGroupingStmt extends YangStatement {
  constructor(init: Partial<YangGroupingStmt> = {}) {
    super(init);
    this.keyword = "grouping";
  }
}

export class YangUsesStmt extends YangStatementWithWhen {
  grouping_name: string;
  refines: YangRefineStmt[];

  constructor(init: Partial<YangUsesStmt> = {}) {
    super(init);
    this.keyword = "uses";
    this.grouping_name = init.grouping_name ?? "";
    this.refines = init.refines ?? [];
  }

  override get_schema_node(): string | undefined {
    return undefined;
  }
}

export class YangAugmentStmt extends YangStatementWithWhen {
  augment_path: string;

  constructor(init: Partial<YangAugmentStmt> = {}) {
    super(init);
    this.keyword = "augment";
    this.augment_path = init.augment_path ?? "";
  }

  override get_schema_node(): string | undefined {
    return undefined;
  }
}

export class YangRefineStmt extends YangStatementWithMust {
  target_path: string;
  type?: YangTypeStmt;
  min_elements?: number;
  max_elements?: number;
  refined_defaults: unknown[];
  refined_mandatory?: boolean;
  if_features: string[];

  constructor(init: Partial<YangRefineStmt> = {}) {
    super(init);
    this.keyword = "refine";
    this.target_path = init.target_path ?? "";
    this.type = init.type;
    this.min_elements = init.min_elements;
    this.max_elements = init.max_elements;
    this.refined_defaults = init.refined_defaults ?? [];
    this.refined_mandatory = init.refined_mandatory;
    this.if_features = init.if_features ?? [];
  }
}

export class YangChoiceStmt extends YangStatementWithWhen {
  mandatory: boolean;
  cases: YangCaseStmt[];

  constructor(init: Partial<YangChoiceStmt> = {}) {
    super(init);
    this.keyword = "choice";
    this.mandatory = init.mandatory ?? false;
    this.cases = init.cases ?? [];
  }

  override child_names(data: Record<string, unknown>): Set<string> {
    for (const c of this.cases) {
      if (c.statements.some((s) => s.name && s.name in data)) {
        return new Set(c.statements.map((s) => s.name).filter((name): name is string => Boolean(name)));
      }
    }
    return new Set();
  }

  validate_case_unique_child_names(): void {
    const seen = new Map<string, string>();
    for (const c of this.cases) {
      for (const sub of c.statements) {
        const seg = sub.get_schema_node();
        if (!seg) {
          continue;
        }
        if (seen.has(seg)) {
          const prevCase = seen.get(seg);
          throw new YangSemanticError(
            `Choice '${this.name}': schema node '${seg}' appears in case '${prevCase}' and again in case '${c.name}' ` +
              "(RFC 7950: names of nodes in the cases of a choice must be unique)."
          );
        }
        seen.set(seg, c.name);
      }
    }
  }
}

export class YangCaseStmt extends YangStatementWithWhen {
  constructor(init: Partial<YangCaseStmt> = {}) {
    super(init);
    this.keyword = "case";
  }

  override child_names(_data: Record<string, unknown>): Set<string> {
    return new Set(this.statements.map((s) => s.name).filter((name): name is string => Boolean(name)));
  }
}
